walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,8 +40,85 @@
     }
 
-    List<TimeRange> requiredTimes = getGoodTimes(events, attendeesRequired, meetingDuration);
-    return requiredTimes;
+    // optimization algorithm
+    // start with one less than max number of optional attendees, and if no times work it will try
+    // one less
+    int numberOptional = Math.max(0, attendeesOptional.size() - 1);
+
+    List<TimeRange> optimalTime =
+        optimizeTime(numberOptional, events, attendeesRequired, attendeesOptional, meetingDuration);
+    return optimalTime;
+  }
+
+  /**
+   * Function that tests all possible combinations of a given number of optional attendees
+   * (numOptional) and returns the times that work. If no times work, it recursively calls itself
+   * with a smaller number of optional attendees until there are times that work. If times don't
+   * work with any optional attendees, it will return the times for only the required attendees.
+   */
+  private static List<TimeRange> optimizeTime(
+      int numOptional,
+      Collection<Event> events,
+      Collection<String> attendeesRequired,
+      Collection<String> attendeesOptionalCollection,
+      int meetingDuration) {
+
+    // change optional attendee collection into list
+    List<String> attendeesOptional = new ArrayList<String>(attendeesOptionalCollection);
+
+    if (numOptional <= 0) { // no optional attendees work
+      return getGoodTimes(events, attendeesRequired, meetingDuration);
+    }
+
+    // add all possible combinations of size numOptional to allOptionalCombos using helper method
+    // combination()
+    List<List<String>> allOptionalCombos = combination(attendeesOptional, numOptional);
+
+    for (List<String> attendeeOptionalCombo : allOptionalCombos) {
+      List<String> attendeesCombined = new ArrayList<String>();
+      attendeesCombined.addAll(attendeesRequired);
+      attendeesCombined.addAll(attendeeOptionalCombo);
+
+      List<TimeRange> potentialTimes = getGoodTimes(events, attendeesCombined, meetingDuration);
+      if (!potentialTimes.isEmpty()) {
+        return potentialTimes;
+      }
+    }
+
+    return optimizeTime(
+        numOptional - 1, events, attendeesRequired, attendeesOptionalCollection, meetingDuration);
+  }
+
+  /**
+   * Given a List and a size, returns a List of all the possible combinations of size (order within
+   * combinations does not matter)
+   */
+  private static List<List<String>> combination(List<String> attendees, int size) {
+    if (size == 0) {
+      return Collections.singletonList(Collections.<String>emptyList());
+    }
+
+    if (attendees.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    List<List<String>> combination = new ArrayList<List<String>>();
+
+    String actual = attendees.iterator().next();
+
+    List<String> subSet = new ArrayList<String>(attendees);
+    subSet.remove(actual);
+
+    List<List<String>> subSetCombination = combination(subSet, size - 1);
+
+    for (List<String> set : subSetCombination) {
```

Comment:
A comment before this loop (or perhaps for each recursive call on 'combination' would be helpful